lib/proxymode.pm

```diff
@@ -0,0 +1,155 @@
+# SUSE's openQA tests
+#
+# Copyright © 2016 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+package proxymode;
+# Summary: proxymode: The basic lib for using proxy mode to connect or operation with physical machine
+# Maintainer: John <xgwang@suse.com>
+
+use base 'basetest';
+use testapi;
+use strict;
+
+our $SLAVE_SERIALDEV = 'proxyserial';
+our $DEBUG           = 1;
+
+sub restart_host {
+    my ($self, $ipmi_machine, $ipmi_user, $ipmi_pass) = @_;
+
+    $ipmi_pass //= 'ADMIN';
+    $ipmi_user //= 'ADMIN';
+    die "Variable ipmi_machine is invalid in function restart_host!!" unless $ipmi_machine;
+
+    my $ipmitool = "ipmitool -H " . $ipmi_machine . " -U " . $ipmi_user . " -P " . $ipmi_pass . " -I lanplus ";
+
+    select_console 'log-console';
+    script_run($ipmitool . 'chassis power off', 20);
+    while (1) {
+        my $stdout = script_output($ipmitool . 'chassis power status', 20);
+        last if $stdout =~ m/is off/;
+        die "Failure on running IPMITOOL" if $stdout =~ m/Error/;
+        script_run($ipmitool . 'chassis power off', 20);
+        sleep(2);
+    }
+    script_run($ipmitool . 'chassis power on', 20);
+    while (1) {
+        my $ret = script_output($ipmitool . 'chassis power status', 20);
+        last if $ret =~ m/is on/;
+        script_run($ipmitool . 'chassis power on', 20);
+        sleep(2);
+    }
+    wait_idle 10;
+    save_screenshot;
+    select_console 'root-console';
+}
+
+sub connect_slave {
+    my ($self, $ipmi_machine, $ipmi_user, $ipmi_pass) = @_;
+    $ipmi_user //= 'ADMIN';
+    $ipmi_pass //= 'ADMIN';
+    die "Variable ipmi_machine is invalid in function connect_slave!!" unless $ipmi_machine;
+    script_run("clear");
+    type_string("ipmitool -H " . $ipmi_machine . " -U " . $ipmi_user . " -P " . $ipmi_pass . " -I lanplus sol activate", 20);
+    send_key 'ret';
+    send_key 'ret';
+    save_screenshot;
+}
+
+sub check_prompt_for_boot() {
+    my $self    = shift;
+    my $timeout = shift;
```

Comment:
Coding standard and also there is some negligible performance impact where assignment from list is cheaper than shifting over the list.
I read somewhere, I thought it was Modern Perl but can't find it now, that assignment is generally preferred to shifting.